### PR TITLE
add CustomFunctionProvider to ExpressionManager

### PR DESCRIPTION
**Pull request by jsokolowskii**
https://github.com/Activiti/Activiti/issues/4101

Adding custom function providers to expression manager.

### CODE REVIEW

File: activiti-core-common/activiti-expression-language/src/main/java/org/activiti/core/el/CustomFunctionProvider.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.activiti.core.el;
+
+public interface CustomFunctionProvider {
+
+    void addCustomFunctions(ActivitiElContext elContext) throws NoSuchMethodException;
```
Comment (erdemedeiros) [inline activiti-core-common/activiti-expression-language/src/main/java/org/activiti/core/el/CustomFunctionProvider.java]: Why do we need to throw `NoSuchMethodException ` here?

**Review by erdemedeiros — CHANGES_REQUESTED** [review]
Apart from my question related to `NoSuchMethodException` it would also be great to add a test covering this. You could define a test bean implementing a very simple `CustomFunctionProvider ` and evaluate an expression that uses it, for example.

File: activiti-core-common/activiti-expression-language/src/main/java/org/activiti/core/el/CustomFunctionProvider.java (hunk 1 above)
Comment (jsokolowskii) [reply activiti-core-common/activiti-expression-language/src/main/java/org/activiti/core/el/CustomFunctionProvider.java]: It's thrown by `class.getMethod()` which is used to provide the method. I can catch it inside the implementation though.

Comment (jsokolowskii) [reply activiti-core-common/activiti-expression-language/src/main/java/org/activiti/core/el/CustomFunctionProvider.java]: Removed.

**Comment by jsokolowskii** [comment]
> Apart from my question related to `NoSuchMethodException` it would also be great to add a test covering this. You could define a test bean implementing a very simple `CustomFunctionProvider ` and evaluate an expression that uses it, for example.

@erdemedeiros I added a simple test. More robust ones should be added in the future.

**Review by erdemedeiros — APPROVED** [review]
LGTM!

**Pull request merged**